### PR TITLE
Implement Simple Naive Bayesian filter

### PR DESCRIPTION
**Pull request by vmax**

### CODE REVIEW

File: projects/noloan/app/filter/bayes/BayesFilter.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        WordInfo wordInfo = message.get(i);
+        positive_probabilities_product *= wordInfo.getSpamProbability();
+        negative_probabilities_product *= (1.0f - wordInfo.getSpamProbability());
+      }
+      double probabilities_products_sum = positive_probabilities_product + negative_probabilities_product;
+      return positive_probabilities_product / probabilities_products_sum;
+    }
+  }
+
+  static class LogProbabilityStrategy implements BayesianProbabilityCalculationStrategy {
```
Comment (oferb) [inline projects/noloan/app/filter/bayes/BayesFilter.java]: Let's choose what we think is the best implementation - simple or log, and make is just a method in BayesFilter.

File: projects/noloan/app/filter/bayes/BayesFilter.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
+package projects.noloan.app.filter.bayes;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import projects.noloan.app.filter.SpamFilterAlgorithm;
+import projects.noloan.app.filter.bayes.Protos.*;
```
Comment (oferb) [inline projects/noloan/app/filter/bayes/BayesFilter.java]: Let's avoid the * import.

File: projects/noloan/app/filter/bayes/BayesFilter.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+   * calculate spam/ham rate and spam probability
+   */
+  private WordInfo wordInfoWithProbability(WordInfo wordInfo) {
+    float spamRate, hamRate, spamProbability = 0;
+
+    spamRate = wordInfo.getSpamCount() / (float) totalSpamCount;
+    hamRate = wordInfo.getHamCount() / (float) totalHamCount;
+    spamProbability = spamRate / (spamRate + hamRate);
+
+    if (spamProbability < 0.01f) {
```
Comment (oferb) [inline projects/noloan/app/filter/bayes/BayesFilter.java]: Let's make these private static floats.

File: projects/noloan/app/filter/bayes/BayesFilter.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    return wordInfo
+            .toBuilder()
+            .setSpamRate(spamRate)
+            .setHamRate(hamRate)
+            .setSpamProbability(spamProbability)
+            .build();
+  }
+
+  private boolean calculateBayesianProbability(ArrayList<WordInfo> sms) {
+    return strategy.probability(sms) > 0.9f;
```
Comment (oferb) [inline projects/noloan/app/filter/bayes/BayesFilter.java]: Let's make this a private static float.

File: projects/noloan/app/filter/bayes/BayesFilter.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
+package projects.noloan.app.filter.bayes;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import projects.noloan.app.filter.SpamFilterAlgorithm;
+import projects.noloan.app.filter.bayes.Protos.*;
+
+public class BayesFilter implements SpamFilterAlgorithm {
```
Comment (oferb) [inline projects/noloan/app/filter/bayes/BayesFilter.java]: I don't think we need an interface. We have just one algorithm at this point.

File: projects/noloan/app/filter/bayes/BayesFilter.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+      double nu = message
+              .stream()
+              .mapToDouble(word ->
+                      Math.log(1 - word.getSpamProbability()) -
+                              Math.log(word.getSpamProbability())).sum();
+      return 1/(1 + Math.exp(nu));
+    }
+  }
+
+  public boolean lineIsSpam(String line) {
```
Comment (oferb) [inline projects/noloan/app/filter/bayes/BayesFilter.java]: How about we return the probability and let the caller decide?

Comment (oferb) [inline projects/noloan/app/filter/bayes/BayesFilter.java]: s/lineIsSpam/isSpam

Comment (oferb) [inline projects/noloan/app/filter/bayes/BayesFilter.java]: s/line/text
or
s/line/message

File: projects/noloan/app/filter/SpamFilter.java [hunk 7]
```diff
@@ -4,20 +4,32 @@
 
 import com.google.common.io.Resources;
 import com.google.common.collect.ImmutableList;
+
```
Comment (oferb) [inline projects/noloan/app/filter/SpamFilter.java]: How about if the spam filter adds the probability to the SmsMessage, for debugging?

File: projects/noloan/app/filter/bayes/BayesFilter.java (hunk 6 above)
Comment (vmax) [reply projects/noloan/app/filter/bayes/BayesFilter.java]: @oferb I view `BayesFilter` as a black box which gives binary answer — either it thinks a message is spam or it is not

File: projects/noloan/app/filter/SpamFilter.java (hunk 7 above)
Comment (vmax) [reply projects/noloan/app/filter/SpamFilter.java]: @oferb won't it break backwards compat with `.prototxt` we have?

Comment (vmax) [reply projects/noloan/app/filter/SpamFilter.java]: @oferb we could `Log` it instead

File: projects/noloan/app/filter/SpamFilterAlgorithm.java [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package projects.noloan.app.filter;
```
Comment (oferb) [inline projects/noloan/app/filter/SpamFilterAlgorithm.java]: I don't think we'll need 2 algorithms at the same time, we'll just evolve the one we have.
So no need for an interface.

**Pull request merged**